File: src/Python/101-200/155.MinStack.py

class MinStack:
    def __init__(self):
        """
        initialize your data structure here.
        """
        self.arr = []
        self.min = 214748364

    def push(self, x):
        """
        :type x: int
        :rtype: void
        """
        if x <= self.min:
            self.arr.append(self.min)
            self.min = x     
        self.arr.append(x)
        
    def pop(self):
        """
        :rtype: void
        """
        element = self.arr.pop()
        if element == self.min:
            self.min = self.arr.pop()
               
    def top(self):
        """
        :rtype: int
        """
        return self.arr[-1]

    def getMin(self):
        """
        :rtype: int
        """
        return self.min

if __name__ == '__main__':
    stack = MinStack()
    stack.push(-2)
    stack.push(0)
    stack.push(-3)
    print(stack.getMin())
    stack.pop()
    stack.top()
    print(stack.getMin())
else:
    pass

File: src/CSharp/101-200/155.MinStack.cs
using System.Collections.Generic;
using System.Linq;

namespace PeefyLeetCode.MinStack
{
    public class MinStack
    {
        int min = int.MaxValue;
        Stack<int> arr = new Stack<int>();
        public MinStack()
        {

        }

        public void Push(int x)
        {
            if (x <= min)
            {
                arr.Push(min);
                min = x;
            }
            arr.Push(x);
        }

        public void Pop()
        {
            var x = arr.Pop();
            if (x == min)
                min = arr.Pop();
        }

        public int Top()
        {
            return arr.Peek();
        }

        public int GetMin()
        {
            return min;
        }
    }
}